src/EditorFeatures/Core.Wpf/InlineParameterNameHints/InlineParameterNameHintsTagger.cs

```diff
@@ -19,56 +23,95 @@ internal sealed class InlineParameterNameHintsTagger : ITagger<IntraTextAdornmen
         {
+            _cache = new List<ITagSpan<IntraTextAdornmentTag>>();
+            _textView = textView;
             _buffer = buffer;
             _tagAggregator = tagAggregator;
+            _formatMap = taggerProvider.ClassificationFormatMapService.GetClassificationFormatMap(textView);
+            _keyword = taggerProvider.ClassificationTypeRegistryService.GetClassificationType(InlineParameterNameHintsTag.TagId);
+            _formatMap.ClassificationFormatMappingChanged += this.OnClassificationFormatMappingChanged;
             _tagAggregator.TagsChanged += OnTagAggregatorTagsChanged;
         }
 
+        private void OnClassificationFormatMappingChanged(object sender, EventArgs e)
+        {
+            if (_format != null)
+            {
+                _format = null;
+                _cache.Clear();
+            }
+        }
+
         private void OnTagAggregatorTagsChanged(object sender, TagsChangedEventArgs e)
         {
+            _cache.Clear();
             var spans = e.Span.GetSpans(_buffer);
             foreach (var span in spans)
             {
                 TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(span));
             }
         }
 
+        private TextFormattingRunProperties Format
+        {
+            get
+            {
+                _format ??= _formatMap.GetTextProperties(_keyword);
+
+                return _format;
+            }
+        }
+
         public IEnumerable<ITagSpan<IntraTextAdornmentTag>> GetTags(NormalizedSnapshotSpanCollection spans)
         {
-            if (spans.Count <= 0)
+            if (spans.Count == 0)
             {
                 return Array.Empty<ITagSpan<IntraTextAdornmentTag>>();
             }
 
-            var tagsList = new List<ITagSpan<IntraTextAdornmentTag>>();
-            var dataTags = _tagAggregator.GetTags(spans);
             var snapshot = spans[0].Snapshot;
-            foreach (var tag in dataTags)
+            if (_cache.Count == 0 || snapshot != _cacheSnapshot)
             {
-                // Gets the associated span from the snapshot span and creates the IntraTextAdornmentTag from the data
-                // tags. Only dealing with the dataTagSpans if the count is 1 because we do not see a multi-buffer case
-                // occuring 
-                var dataTagSpans = tag.Span.GetSpans(snapshot);
-                var textTag = tag.Tag;
-                if (dataTagSpans.Count == 1)
+                // Calculate UI elements
+                _cache.Clear();
+                _cacheSnapshot = snapshot;
+                var fullSpan = new SnapshotSpan(snapshot, 0, snapshot.Length);
+                var requestSpans = new NormalizedSnapshotSpanCollection(fullSpan);
+                var dataTags = _tagAggregator.GetTags(requestSpans);
```

Comment:
I recognize we're still investigating why this cache fixes the hang issue but if we merge this in let's file a tracking bug to not forget to remove this if we get distracted by other work. Making UI for all the tags on the UI thread even if they're not visible isn't great, but works for the short term.

Reply:
@jasonmalinowski  is it possible the following is happening:

1. we compute tags
2. we report tags.
3. editor calls into us to get tags and then displays them.
4. displaying the tag causes the line to change (due to the adornment)
5. we're listening to line changes and we kick off anotehr computation

rinse repeat?

Reply:
@CyrusNajmabadi We figure it's something like that, yeah. Displaying the tag will cause the editor to have to lay out the text again but that won't create a new snapshot or reprocess it. We chatted with the editor team and they were aware of one potential issue; the editor calls for tags twice during the process of laying everything out; and if between those two times you are _replacing_ text with an adornment (which the editor supports, but we're not doing here) _and_ you changed the tags between those two calls you might end up with extra layouts or bit of a cycle until things stabilize. But we're not using that editor feature, and our async tagger correctly raises tags changed on the UI thread, and only then reports new tags, so it looks like everything is fine there.

Reply:
The samples that the editor team have often use a cache of the adornment tags in general and we could bring that code into here; our original belief was because our async tagger wasn't doing the bad things the editor knew to avoid we thought we were good and didn't need a caching layer atop the async tagger's caching layer. @akhera99 is still investigating further and trying to narrow it down.